app/domain/authentication/authn_azure/validate_application_identity.rb

```diff
@@ -9,119 +9,158 @@ module AuthnAzure
-      def validate_token_identity_matches_annotations
-        application_identity.constraints.each do |constraint|
-          annotation_type  = constraint[0].to_s
-          annotation_value = constraint[1]
-          unless annotation_value == @token_identity[annotation_type.to_sym]
-            raise Err::InvalidApplicationIdentity, annotation_type_constraint(annotation_type)
-          end
+      # xms_mirid is a term in Azure to define a claim that describes the resource that holds the encoding of the instance's
+      # among other details the subscription_id, resource group, and provider identity needed for authorization.
+      # xms_mirid is one of the fields in the JWT token. This function will extract the relevant information from
+      # xms_mirid claim and populate a representative hash with the appropriate fields.
+      def extract_resources_from_token
+        parse_xms_mirid
+        @resources_from_token = [
+          AzureResource.new(
+            resource_type: "subscription-id",
+            resource_name: xms_mirid.subscriptions
+          ),
+          AzureResource.new(
+            resource_type: "resource-group",
+            resource_name: xms_mirid.resource_groups
+          )
+        ]
+
+        # determine which identity is provided in the token. If the token is
+        # issued to a user-assigned identity then we take the identity name.
+        # If the token is issued to a system-assigned identity then we take the
+        # Object ID of the token.
+        if xms_mirid.providers.include? "Microsoft.ManagedIdentity"
+          @resources_from_token.push(
+            AzureResource.new(
+              resource_type: "user-assigned-identity",
+              resource_name: xms_mirid.providers.last
+            )
+          )
+        else
+          @resources_from_token.push(
+            AzureResource.new(
+              resource_type: "system-assigned-identity",
+              resource_name: @oid_token_field
+            )
+          )
         end
-        @logger.debug(Log::ValidatedApplicationIdentity.new)
+        @logger.debug(Log::ExtractedApplicationIdentityFromToken.new)
       end
 
-      def annotation_type_constraints constraints
-        constraints.map { |constraint| annotation_type_constraint(constraint) }
+      def parse_xms_mirid
+        xms_mirid
+      end
+
+      def xms_mirid
+        @xms_mirid ||= XmsMirid.new(@xms_mirid_token_field)
       end
 
-      # converts the constraint to be in annotation style (e.g resource-group
-      # instead of resource_group) to enhance supportability
-      def annotation_type_constraint constraint
-        constraint.to_s.tr('_', '-')
+      def validate_application_identity_matches_request
+        extract_resources_from_token
+
+        application_identity.resources.each do |resource_from_role|
+          @resources_from_token.each do |resource_from_token|
+            if resource_from_token.resource_type == resource_from_role.resource_type
```

Comment:
why do we have an if unless here? 
`if resource_from_token.resource_type == resource_from_role.resource_type` do logging?

If so, I don't think that is very clear

Reply:
we want to compare between the values of the resource object of the same resource type. For example, if the subscription id is `sub_1` and the resource_group is `rg_1`, i would like to compare `sub_1` with the subscription id in the token, and not with the resource_group in the token.

So first, we find the corresponding resource type and then compare between its value. 

I'll change it from:
```
if resource_from_token.resource_type == resource_from_role.resource_type
              unless resource_from_token.resource_name == resource_from_role.resource_name
                raise Err::InvalidApplicationIdentity, resource_from_role.resource_type
              end
            end
```

to:
```
if resource_from_token.resource_type == resource_from_role.resource_type && 
              resource_from_token.resource_name != resource_from_role.resource_name
                raise Err::InvalidApplicationIdentity, resource_from_role.resource_type
              end
            end
```.

Is this better?

Reply:
much!